### PR TITLE
more POSIXish MEMFS: update timestamp in parent on file create/deletion/rename

### PR DESCRIPTION
**Pull request by nopid**
tl;dr Modify MEMFS behavior to look more POSIXish: the timestamp of a directory is modified each time a file is created or deleted inside the directory.

This PR comes from the resolution of [caching behavior issues in pyodide](https://github.com/iodide-project/pyodide/issues/737).

On a modern Unix-like system, each file (including directories) has up to 4 timestamps: creation, change, modification and access date and time. Three of them are available through the standard POSIX stat() function call:
```C
    time_t    st_atime;   /* time of last access */
    time_t    st_mtime;   /* time of last modification */
    time_t    st_ctime;   /* time of last status change */
```

In MEMFS there is a single timestamp attribute per node:
```javascript
      node.timestamp = Date.now();
```

In its sourcecode it is accessed only during the following operations:
 1. [`createNode`](https://github.com/emscripten-core/emscripten/blob/af9a5a703464e93e2ae9182e28305c42fef97377/src/library_memfs.js#L88) when it is initialized with `Date.now()`
 2. [`getattr`](https://github.com/emscripten-core/emscripten/blob/af9a5a703464e93e2ae9182e28305c42fef97377/src/library_memfs.js#L181-L183) where the value is read to fill the fields `attr.atime`, `attr.mtime` and `attr.ctime`
 3. [`setattr`](https://github.com/emscripten-core/emscripten/blob/af9a5a703464e93e2ae9182e28305c42fef97377/src/library_memfs.js#L194-L195) where the value is replaced by the field `attr.timestamp` if filled
 4. [`write`](https://github.com/emscripten-core/emscripten/blob/af9a5a703464e93e2ae9182e28305c42fef97377/src/library_memfs.js#L298) where it is updated to `Date.now()` each time some data is written to a file (excluding directories and links that do not support the write operation)

Thus the current semantics of `timestamp` in MEMFS is more like a last modification timestamp for files and a creation timestamp for directories.

My proposal is to modify `createNode` to also update the `timestamp` of the parent directory so that the virtual filesystem behaves like the modification timestamp of a POSIX.1 ([open](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html), [unlink](https://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html)) compatible filesystem:

> If O_CREAT is set and the file did not previously exist, upon successful completion, open() shall mark for update the last data access, last data modification, and last file status change timestamps of the file and the last data modification and last file status change timestamps of the parent directory.

> Upon successful completion, unlink() shall mark for update the last data modification and last file status change timestamps of the parent directory. Also, if the file's link count is not 0, the last file status change timestamp of the file shall be marked for update.

The change is very minimal, no space overhead and very little time overhead.

Please notice that this is very different from implementing a full POSIX filesystem (space and time overhead) or implementing a separation of access and modification timestamps as discussed in #11454.

### CODE REVIEW

**Review by sbc100 — APPROVED** [review]
Looks great.  Thanks!

Can we write a test for this?

**Comment by nopid** [comment]
Thanks @sbc100. Something like this would test with assertions that directory are properly updated on file creation and unlink. If this is the kind of test you have in mind, what would be the proper way to properly include it into emscripten tests?

```C
#include <assert.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/stat.h>

void mysleep(int delta)
{
    time_t end = time(NULL) + delta;
    while (time(NULL) < end);
}

time_t getmtime(char *fname)
{
    struct stat buf;
    stat(fname, &buf);
    return buf.st_mtime;
}

void writefile(char *fname, char *content)
{
    int fd = open(fname, O_CREAT | O_TRUNC | O_WRONLY, 0777);
    write(fd, content, strlen(content));
    close(fd);
}

int main(int argc, char *argv[])
{
    char tmpdir[64] = "/tmp/tmpXXXXXXX";
    char fname[64];
    time_t t0, t1, t2, t3;
    mkdtemp(tmpdir);
    strcpy(fname, tmpdir);
    strcat(fname, "/foo.py");
    t0 = getmtime(tmpdir);
    mysleep(1);
    writefile(fname, "bar = 54\n");
    t1 = getmtime(tmpdir);
    t2 = getmtime(fname);
    assert(t1 > t0);
    assert(t1 == t2);
    mysleep(1);
    unlink(fname);
    t3 = getmtime(tmpdir);
    assert(t3 > t1);
    rmdir(tmpdir);
}
```

**Comment by sbc100** [comment]
I would modify the existing `tests/stat/test_stat.c`.    You can run it like this: "./tests/runner.py wasm0.test_stat`

**Comment by nopid** [comment]
Thanks @sbc100, test added to `tests/stat/test_stat.c`.

File: tests/stat/test_stat.c [hunk 1]
```diff
@@ -29,6 +30,11 @@ void create_file(const char *path, const char *buffer, int mode) {
   close(fd);
 }
 
+void active_sleep(int delta) {
+    time_t end = time(NULL) + delta;
+    while (time(NULL) < end);
+}
```
Comment (sbc100) [inline tests/stat/test_stat.c:36]: If possible its best if we can avoid buys waiting in tests.

Can you perhaps explicitly set the mtime of a directory to be in the past to ensure it gets correctly updated, rather than sleeping/waiting?

Comment (nopid) [reply tests/stat/test_stat.c]: sure, did it using the same technic as the other tests in `test_stat.c`

File: tests/stat/test_stat.c [hunk 2]
```diff
@@ -161,6 +164,17 @@ void test() {
...
+  // create and unlink files inside a directory and check that mtime updates
+  mkdir("folder/subdir", 0777);
+  utime("folder/subdir", &t);
+  create_file("folder/subdir/file", "abcdef", 0777);
+  err = stat("folder/subdir", &s);
+  assert(s.st_mtime != 1200000000);
+  utime("folder/subdir", &t);
+  unlink("folder/subdir/file");
+  err = stat("folder/subdir", &s);
+  assert(s.st_mtime != 1200000000);
```
Comment (sbc100) [inline tests/stat/test_stat.c:176]: Should we assert that its near to the current time?   Maybe that is too timing sensitive?   How about at least replacing != with strictly greater than?

lgtm either way... let me know if you'd prefer to land this as is.

**Review by sbc100 — APPROVED** [review]

File: tests/stat/test_stat.c (hunk 2 above)
Comment (nopid) [reply tests/stat/test_stat.c:176]: Being different from 2008-01-10T22:20:00 should be enough.
